python_ta/checkers/redundant_or_impossible_condition_checker.py

```diff
@@ -0,0 +1,117 @@
+"""
+Check for redundant If or While conditions in functions based on z3 constraints
+"""
+
+from typing import Any, Union
+
+try:
+    import z3
+
+    z3_dependency_available = True
+except ImportError:
+    z3 = Any
+    z3_dependency_available = False
+
+from astroid import nodes
+from pylint.checkers import BaseChecker
+from pylint.checkers.utils import only_required_for_messages
+from pylint.lint import PyLinter
+
+
+class RedundantOrImpossibleConditionChecker(BaseChecker):
+    name = "redundant-condition"
+    msgs = {
+        "R9900": (
+            """This conditional statement is already true given the preceding code;
+             consider removing redundant check.""",
+            "redundant-condition",
+            "Used when an If or While statement is tautologically true based on preceding Z3 constraints."
+            " Enabled when Z3 option is on.",
+        ),
+        "R9901": (
+            """This conditional statement is never true given the preceding code.""",
+            "impossible-condition",
+            "Used when an If or While statement contradicts preceding Z3 constraints. Enable when Z3 option is on.",
+        ),
+    }
+    options = (
+        (
+            "z3",
+            {
+                "default": False,
+                "type": "yn",
+                "metavar": "<y or n>",
+                "help": "Use Z3 to perform logical feasibility analysis in program control flow.",
+            },
+        ),
+    )
+
+    def __init__(self, linter=None) -> None:
+        super().__init__(linter=linter)
+
+    @only_required_for_messages("redundant-condition", "impossible-condition")
+    def visit_if(self, node: nodes.If) -> None:
+        """Visit if statement"""
+        self._check_condition(node)
+        # handle else branch
+
+    @only_required_for_messages("redundant-condition", "impossible-condition")
+    def visit_while(self, node: nodes.While) -> None:
+        """Visit while statement"""
+        self._check_condition(node)
+
+    def _check_condition(self, node: Union[nodes.If, nodes.While]) -> None:
+        """A condition statement is redundant if for every feasible execution path
+        leading to the node, the condition must be True by precedent constraints.
+        """
+        if (
+            not hasattr(node, "cfg_block")
+            or not z3_dependency_available
+            or not self.linter.config.z3
+        ):
+            return
+
+        node_block = node.cfg_block
+
+        if not node_block.is_feasible or node_block.z3_constraint is None:
+            return
+
+        for pred in node_block.predecessors:
```

Comment:
only include feasible edges here; this also removes the need for a `node_block.is_feasible` check above, as if there are no feasible predecessor edges the loop won't run at all